### PR TITLE
Use c2h::vectors in all non-example CUB tests

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Two CUB tests, `catch2_test_device_select_if_vsmem.cu` and `test_device_spmv.cu`, where not yet using `c2h::[host|device]_vector`. This PR changes that.

### CODE REVIEW

**Review by alliepiper — CHANGES_REQUESTED** [review]
This partially addresses the issue that `c2h` vectors fix.

Some of our supported platforms behave very badly when requesting an allocation that exceeds the amount of free memory, and will just hang forever instead of [throwing, swapping, erroring, crashing, or literally doing *anything* sensible]. These vectors use a checked allocator to pre-check the amount of free mem on these platforms and throw bad_alloc if needed, *before* making the actual memory request.

`c2h::device_vector` and `c2h::host_vector` both implement these checks, so this PR covers the vector allocs.

However, many Thrust algorithms also allocate temporary storage, and these allocations need to be pre-checked as well. This is accomplished by passing `c2h::device_policy` as the execution policy for all Thrust algorithms that run on device. This policy provides the same checked allocator for any temp storage allocations.

So basically, all thrust algorithms that run on device need to use `c2h::device_policy` in our tests.

The exception is `thrust::copy`, since specifying a policy for cross-system copies will break things. But `thrust::copy` won't allocate temporary storage, so it's fine to omit the policy here.

File: cub/test/test_device_spmv.cu [hunk 1]
```diff
@@ -121,7 +123,8 @@ struct csr_matrix
 
   void finalize()
   {
-    thrust::exclusive_scan(m_row_offsets.cbegin(), m_row_offsets.cend(), m_row_offsets.begin());
+    using exec_policy_t = cub::detail::conditional_t<HostStorage, decltype(thrust::host), c2h::device_policy_t>;
+    thrust::exclusive_scan(exec_policy_t{}, m_row_offsets.cbegin(), m_row_offsets.cend(), m_row_offsets.begin());
```
Comment (alliepiper) [inline cub/test/test_device_spmv.cu:127]: I'm not sure how this is compiling 😕

`c2h::device_policy_t` is an alias for

```
thrust::detail::execute_with_allocator<
  c2h::checked_cuda_allocator<char>, 
  thrust::cuda_cub::execute_on_stream_base
>
```

and `execute_with_allocator` [doesn't have a default ctor](https://github.com/NVIDIA/cccl/blob/main/thrust/thrust/detail/execute_with_allocator_fwd.h#L46-L54).

Maybe I'm missing something, but it seems like this should be:

```suggestion
    _CCCL_IF_CONSTEXPR(HostStorage)
    {
      thrust::exclusive_scan(thrust::host, m_row_offsets.cbegin(), m_row_offsets.cend(), m_row_offsets.begin());
    }
    else
    {
      thrust::exclusive_scan(c2h::device_policy, m_row_offsets.cbegin(), m_row_offsets.cend(), m_row_offsets.begin());
    }   
```

**Review by alliepiper — APPROVED** [review]
Thanks for the updates!

I'm confused about one part of this, see inline comment. If we can figure out why this is compiling and prove that this syntax is indeed valid, we can leave it as is, otherwise we should avoid default-constructing `c2h::device_policy_t`.

File: cub/test/test_device_spmv.cu (hunk 1 above)
Comment (alliepiper) [reply cub/test/test_device_spmv.cu:127]: Maybe `HostStorage` is always true?

Comment (alliepiper) [reply cub/test/test_device_spmv.cu:127]: Nope, `device_csr_matrix` is used in the test.

Huh.

Comment (bernhardmgruber) [reply cub/test/test_device_spmv.cu:127]: It seems `csr_matrix::finalize()` is never called when `HostStorage` is false. I am not sure this is legal C++, since `finalize` does not have any template parameters, so I think it must be instantiated (and checked) when `csr_matrix` is instantiated.
